cpp/src/groupby/sort/sort_helper.cu

```diff
@@ -0,0 +1,312 @@
+                        group_labels.begin());
+
+  return group_labels;
+}
+
+column_view helper::unsorted_keys_labels(cudaStream_t stream) {
+  if (_unsorted_keys_labels)
+    return _unsorted_keys_labels->view();
+
+  column_ptr temp_labels = make_numeric_column(
+                              data_type(type_to_id<size_type>()),
+                              _keys.num_rows(), 
+                              mask_state::ALL_NULL, stream);
+  
+  auto group_labels_view = cudf::column_view(
+                              data_type(type_to_id<size_type>()),
+                              group_labels().size(),
+                              group_labels().data().get());
+  
+  auto scatter_map = cudf::experimental::detail::slice(
+    key_sort_order(), 0, num_keys(stream));
+
+  std::unique_ptr<table> t_unsorted_keys_labels = 
+    cudf::experimental::detail::scatter(
+      table_view({group_labels_view}), scatter_map, 
+      table_view({temp_labels->view()}),
+      false, rmm::mr::get_default_resource(), stream);
+
+  _unsorted_keys_labels = std::move(t_unsorted_keys_labels->release()[0]);
+
+  return _unsorted_keys_labels->view();
+}
+
+column_view helper::keys_bitmask_column(cudaStream_t stream) {
+  if (_keys_bitmask_column)
+    return _keys_bitmask_column->view();
+
+  auto row_bitmask = bitmask_and(_keys, rmm::mr::get_default_resource(), stream);
+
+  _keys_bitmask_column = make_numeric_column(
+    data_type(type_id::INT8), _keys.num_rows(), std::move(row_bitmask),
+    cudf::UNKNOWN_NULL_COUNT, stream);
+
+  using T = id_to_type<type_id::INT8>;
+  thrust::fill(rmm::exec_policy(stream)->on(stream),
+    _keys_bitmask_column->mutable_view().begin<T>(),
+    _keys_bitmask_column->mutable_view().end<T>(),
+    0);
+
+  return _keys_bitmask_column->view();
+}
+
+helper::column_ptr helper::sorted_values(column_view const& values, 
+  rmm::mr::device_memory_resource* mr,
+  cudaStream_t stream)
+{
+  column_ptr values_sort_order = cudf::experimental::detail::sorted_order(
+    table_view({unsorted_keys_labels(), values}), {},
+    std::vector<null_order>(2, null_order::AFTER), mr, stream);
+
+  // Zero-copy slice this sort order so that its new size is num_keys()
+  column_view gather_map = cudf::experimental::detail::slice(
+    values_sort_order->view(), 0, num_keys(stream) );
+
+  auto sorted_values_table = cudf::experimental::detail::gather(
+    table_view({values}), gather_map, false, false, false, mr, stream);
+
+  return std::move(sorted_values_table->release()[0]);
+}
+
+helper::column_ptr helper::grouped_values(column_view const& values, 
+  rmm::mr::device_memory_resource* mr,
+  cudaStream_t stream)
+{
+  auto gather_map = cudf::experimental::detail::slice(
+    key_sort_order(), 0, num_keys(stream));
+
+  auto grouped_values_table = cudf::experimental::detail::gather(
+    table_view({values}), gather_map, false, false, false, mr, stream);
+
+  return std::move(grouped_values_table->release()[0]);
```

Comment:
would that be better?

Reply:
I think it would be slightly clearer. 

`std::move(grouped_values_table->get_column(0))` tells you that you are moving the contents from the first column.

`return std::move(grouped_values_table->release()[0]);` this requires you to know that `release()` returns a vector and accessing the 0th element (which is the first column). Plus, using the `operator[]` on the temporary vector returned from `release()` may be confusing to some.

Reply:
BTW, `release()` gives a vector of `unique_ptr<column>` whereas `get_column(int)` returns a column by ref. So it's going to be 
```c++
std::make_unique<column>(std::move(grouped_values_table->get_column(0)))
```